pkgpanda/util.py

```diff
@@ -264,13 +265,32 @@ def make_file(name):
 
 
 def write_json(filename, data):
-    with open(filename, "w+") as f:
-        return json.dump(data, f, **json_prettyprint_args)
+    dumped_json = json.dumps(data, **json_prettyprint_args)
+    write_string(filename, dumped_json)
 
 
 def write_string(filename, data):
-    with open(filename, "w+") as f:
-        return f.write(data)
+    """
+    Write a string to a file.
+    Overwrite any data in that file.
+
+    We use an atomic write practice of creating a temporary file and then
+    moving that temporary file to the given ``filename``. This prevents race
+    conditions such as the file being read by another process after it is
+    opened here but not yet written to.
+
+    It also prevents us from creating or truncating a file before we fail to
+    write data to it because of low disk space.
+    """
+    temporary_filename = '{filename}-{random}.tmp'.format(
+        filename=filename,
+        random=uuid.uuid4().hex,
+    )
```

Comment:
Why does this not use the well-tested cross-platform Python library https://docs.python.org/2/library/tempfile.html#tempfile.mkstemp ?

Reply:
And of course, I should point you to the Python 3 version rather than the Python 2 version: https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp

Reply:
I originally used `mkstemp` and got an error because the created file did not exist on the same partition as the `replace` destination. However, I guess I can use the `prefix` argument.